backend/cn/lib/builtins.ml

```diff
@@ -36,12 +41,52 @@ let mk_arg3_err mk args loc =
 
 let mk_arg3 mk = mk_arg3_err (fun tup loc -> return (mk tup loc))
 
-let mk_arg5 mk args loc =
-  match args with
-  | [ a; b; c; d; e ] -> return (mk (a, b, c, d, e) loc)
-  | xs -> fail { loc; msg = Number_arguments { has = List.length xs; expect = 5 } }
+let binop op (lhs, rhs) =
+  Some
+    CF.Cn.(
+      CNExpr
+        ( builtin_location,
+          CNExpr_binop
+            ( op,
+              CNExpr (builtin_location, CNExpr_var lhs),
+              CNExpr (builtin_location, CNExpr_var rhs) ) ))
+
+
+let mk_builtin_arg1 name t1 mk return =
+  let arg1 = Sym.fresh_named "arg1" in
+  let sym = Sym.fresh_named name in
+  ( name,
+    sym,
+    CF.Cn.
+      { cn_func_magic_loc = builtin_location;
+        cn_func_loc = builtin_location;
+        cn_func_name = sym;
+        cn_func_attrs = [];
+        cn_func_args = [ (arg1, t1) ];
+        cn_func_body = mk arg1;
+        cn_func_return_bty = return
+      } )
+
+
+let mk_builtin_arg2 name (t1, t2) mk return =
+  let arg1 = Sym.fresh_named "arg1" in
+  let arg2 = Sym.fresh_named "arg2" in
+  let sym = Sym.fresh_named name in
+  ( name,
+    sym,
+    CF.Cn.
+      { cn_func_magic_loc = builtin_location;
+        cn_func_loc = builtin_location;
+        cn_func_name = sym;
+        cn_func_attrs = [];
+        cn_func_args = [ (arg1, t1); (arg2, t2) ];
+        cn_func_body = mk (arg1, arg2);
+        cn_func_return_bty = return
+      } )
 
 
+let arith_binop op (it, it') loc = IT (Binop (op, it, it'), bt it, loc)
+
 let min_bits_def (sign, n) =
```

Comment:
A lot of these functions remain to be translated into the new `builtin_fn` type, but not all of them seem possible to do at this time, for a couple reasons:

1) Some of the binops supported in the `IT.binop` type don't have equivalents in the `CN.binop` type (e.g. `ShiftLeft` and `ShiftRight`), so it is unclear to me how to write definitions for these.

2) Some of the binops that are supported (e.g. `BW_And` and `BW_Or`) appear to be functionally polymorphic, in that they operate on unsigned integers of multiple different sizes. Is it possible to specify a function definition that is polymorphic in this way in the CN intermediate language?

Reply:
I reverted back to the original plan which addressed issue 1 here, but issue 2 here remains an issue: some of these builtins operate over all types of `int`s and `uint`s, rather than one specific input type. For something like `ptr_eq` we can just create a function with two `pointer` arguments whose body is an index term comparing their value, but for `mod`, for example, it does not seem possible to define a concrete `LogicalFunction.definition` that permits multiple types of arguments (e.g. a `u32` and a `u64` both).

Reply:
In that case, can you convert the ones you can and leave a comment that we need basetype polymorphism for the ones leftover. It's a bit ugly and unfortunate, but something is better than nothing here.